### PR TITLE
network: batch transactions on broadcast

### LINKED ISSUE #665: TX broadcast batching
Sending one tx hash per `inv` message adds about 70% wrapper overhead to this tx hash. It also makes the receiving node request transactions one-by-one sending respective `getdata` messages with the same overhead. It would be nice to make this overhead less than 5% under load, but at the same time keeping the overall system responsiveness (not delaying sends a lot).

We can batch tx broadcasts, queuing up to 32 transactions in one inv with a timeout of 50 ms, it's about 640 tx/s rate which seems to be reasonable in that higher load will always be fully batched and with lower load we'll not wait a lot, 50 ms is not noticeable for users (and that's maximum delay).

Supposedly this move will improve TPS.

### PR DESCRIPTION
**Pull request by fyrchik**
Closes #665 .

### CODE REVIEW

**Comment by fyrchik** [comment]
Some results on benchmarking:
I have tried `neo-bench` with different settings. Everytime RPS was higher after this patch.

Noticeable changes: after increasing mempool size to 500k, this bench showed interesting results.
`.make/runner.sh -f $(DC_GO_IR) -f $(DC_GO_RPC) -i /dump.txs -d "Go4x1" -m rate -q 1000 -z 5m -t 30s -a go-node:20331`

Before the patch (on the current `master`) amount of tx per block was roughly the same (8k -- 12k) with new blocks accepted ~every 20s.
After the patch amount of tx per block has been steadily increased during the bench up to 30k tx and  30s per block. Because we propose ALL currently verified transactions in a block, this can show that transactions become verified faster.

File: pkg/network/server.go [hunk 1]
```diff
@@ -813,11 +817,66 @@ func (s *Server) RelayTxn(t *transaction.Transaction) RelayReason {
...
+		batchSize = 32
+	)
+
+	txs := make([]util.Uint256, 0, batchSize)
+
+	timer := time.NewTimer(batchTime)
+	defer timer.Stop()
+
+	broadcast := func() {
+		go s.broadcastTxHashes(txs)
```
Comment (roman-khimov) [inline pkg/network/server.go]: That seems to be a little dangerous, have you tried not spawning a goroutine?

File: pkg/network/server.go [hunk 2]
```diff
@@ -813,11 +817,66 @@ func (s *Server) RelayTxn(t *transaction.Transaction) RelayReason {
...
+		default:
+		}
+
+		select {
+		case <-s.quit:
+			return
+		case <-timer.C:
+			if len(txs) > 0 {
+				broadcast()
+			}
```
Comment (roman-khimov) [inline pkg/network/server.go]: So if there are no transactions our timer is stopped until the other branch reaches `broadcast()`?

File: pkg/network/server.go [hunk 3]
```diff
@@ -813,11 +817,66 @@ func (s *Server) RelayTxn(t *transaction.Transaction) RelayReason {
...
+		case tx := <-s.transactions:
+			txs = append(txs, tx.Hash())
```
Comment (roman-khimov) [inline pkg/network/server.go]: We actually can start timer here when `len(txs) == 0` and it's not yet started. No transactions --- no timer. And broadcaster would just stop it, not restarting.

File: pkg/network/server.go [hunk 4]
```diff
@@ -826,11 +830,66 @@ func (s *Server) RelayTxn(t *transaction.Transaction) RelayReason {
...
+
+	for {
+		select {
+		case <-s.quit:
+			return
+		case <-timer.C:
+			if len(txs) > 0 {
+				broadcast()
+			}
+		default:
```
Comment (roman-khimov) [inline pkg/network/server.go]: I don't think we need this select, it's not a problem to miss a timer if there are incoming transactions (we'll send a batch if it's full anyway).

File: pkg/network/server.go [hunk 5]
```diff
@@ -826,11 +830,66 @@ func (s *Server) RelayTxn(t *transaction.Transaction) RelayReason {
...
 
 	// We need to filter out non-relaying nodes, so plain broadcast
 	// functions don't fit here.
 	s.iteratePeersWithSendMsg(msg, Peer.EnqueuePacket, func(p Peer) bool {
 		return p.Handshaked() && p.Version().Relay
 	})
 }
+
+// broadcastTxLoop is a loop for batching and sending
+// transactions hashes in and INV payload.
```
Comment (roman-khimov) [inline pkg/network/server.go]: `s/and/an/` while you're here.

File: pkg/network/server.go (hunk 2 above)
Comment (fyrchik) [reply pkg/network/server.go]: Fixed most of the issues.

File: pkg/network/server.go (hunk 1 above)
Comment (fyrchik) [reply pkg/network/server.go]: Yes. The problem is slow peers can possibly hang all tx broadcasting.

Comment (roman-khimov) [reply pkg/network/server.go]: Yeah, but that's a known different problem, at the moment I fear more of unbalanced pipeline and goroutine leakage.

**Comment by roman-khimov** [comment]
From what I see in my testing of it, RPS is improved and block generation is changed a bit in that it really tends to produce bigger and bigger blocks over time. But at the same time I don't see any clear negative effect on TPS, so even if it just improves RPS it's a good change.

**Pull request merged**